src/core/build_target.go

```diff
@@ -461,6 +461,54 @@ func (target *BuildTarget) AllURLs(state *BuildState) []string {
 	return ret
 }
 
+// resolveDependencies matches up all declared dependencies to the actual build targets.
+// TODO(peterebden,tatskaari): Work out if we really want to have this and how the suite of *Dependencies functions
+//                             below should behave (preferably nicely).
+func (target *BuildTarget) resolveDependencies(graph *BuildGraph, callback func(*BuildTarget) error) error {
+	var g errgroup.Group
+	target.mutex.RLock()
+	for i := range target.dependencies {
+		dep := &target.dependencies[i]
+		if len(dep.deps) > 0 {
+			continue  // already done
+		}
+		g.Go(func() error {
+			if err := target.resolveOneDependency(graph, dep); err != nil {
+				return err
+			}
+			for _, d := range dep.deps {
+				if err := callback(d); err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+	}
+	target.mutex.RUnlock()
+	return g.Wait()
+}
+
+func (target *BuildTarget) resolveOneDependency(graph *BuildGraph, dep *depInfo) error {
+	t := graph.WaitForTarget(dep.declared)
+	if t == nil {
+		return fmt.Errorf("Couldn't find dependency %s", dep.declared)
+	}
+	labels := t.ProvideFor(target)
+	// Small optimisation to avoid re-looking-up the same target again.
+	if len(labels) == 1 && labels[0] == t.Label {
+		dep.deps = []*BuildTarget{t}
+		return nil
+	}
```

Comment:
Does this make a big difference? I guess it might because it avoids many mutex locks on the main target map, and this is the "normal" way to depend on things.

Reply:
Yeah maybe it is a bit premature, but given the aim of the base PR was to reduce time spent on those map lookups it feels relevant.
I think it's also nice to separate the error messages as well.